src/spec.rs

```diff
@@ -1,19 +1,94 @@
 use crate::result::{ZipError, ZipResult};
-use crate::unstable::{LittleEndianReadExt, LittleEndianWriteExt};
+use memchr::memmem::FinderRev;
 use std::borrow::Cow;
 use std::io;
 use std::io::prelude::*;
+use std::mem;
 use std::path::{Component, Path};
 
-pub const LOCAL_FILE_HEADER_SIGNATURE: u32 = 0x04034b50;
-pub const CENTRAL_DIRECTORY_HEADER_SIGNATURE: u32 = 0x02014b50;
-pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06054b50;
-pub const ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06064b50;
-pub(crate) const ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE: u32 = 0x07064b50;
+pub type Magic = u32;
+
+pub const LOCAL_FILE_HEADER_SIGNATURE: Magic = 0x04034b50;
+pub const CENTRAL_DIRECTORY_HEADER_SIGNATURE: Magic = 0x02014b50;
+pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: Magic = 0x06054b50;
+pub const ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE: Magic = 0x06064b50;
+pub(crate) const ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE: Magic = 0x07064b50;
 
 pub const ZIP64_BYTES_THR: u64 = u32::MAX as u64;
 pub const ZIP64_ENTRY_THR: usize = u16::MAX as usize;
 
+pub trait Block: Sized {
+    fn interpret(bytes: Vec<u8>) -> ZipResult<Self>;
+
+    fn parse<T: Read>(reader: &mut T) -> ZipResult<Self> {
+        let mut block = vec![0u8; mem::size_of::<Self>()];
+        reader.read_exact(&mut block)?;
+        Self::interpret(block)
+    }
+
+    fn encode(self) -> Vec<u8>;
```

Comment:
This can return `Box<[u8]>` instead, assuming it rarely needs to be appended to after being returned.

Reply:
Done! I'm also going to add another method to reduce the boilerplate with `mem::transmute()`, as well as adding a unit test for the unsafe byte munging as you requested.

Reply:
Done! See the new `block_serde()` test.